docs/internals/data_file.md

```diff
@@ -0,0 +1,196 @@
+log (WAL). The WAL is a ring buffer with prepares, and represents the logical diff which should be
+applied to the state represented by superblock/grid to get the actual current state of the system.
+WAL inner workings are described in the [VSR documentation](./vsr.md#protocol-normal), but, on a
+high-level, when a replica processes a prepare, the replica:
+
+* writes the prepare to the WAL on disk
+* applies changes from the prepare to the in-memory data structure representing the current state
+* applies changes from the prepare to the pending on-disk state by allocating and writing fresh grid
+  blocks
+
+When enough prepares are received, the superblock is updated to point to the accumulated-so-far new
+disk state.
+
+This covers how the three major areas of the data file -- the write-ahead log, the superblock and
+the grid -- work together to represent abstract persistent logical state.
+
+Concretely, the state of TigerBeetle is a collection (forest) of LSM trees. LSM structure is
+described [in a separate document](./lsm.md), here only high level on-disk layout is discussed.
+
+Each LSM tree stores a set of values. Values are:
+
+* fixed in size,
+* small (hundreds of bytes),
+* sorted by key,
+* which is embedded in the value itself (e.g, an `Account` value uses `timestamp` as a unique key).
+
+To start from the middle, values are arranged in tables on disk. Each table represents a sorted
+array of values and is physically stored in multiple blocks. Specifically:
+
+* A table's data blocks each store a sorted array of values.
+* A table's index block stores pointers to the data blocks, as well as boundary keys.
+
+```zig
+const TableDataBlock = struct {
+    values_sorted: [value_count_max]Value,
+};
+
+const TableIndexBlock = struct {
+    data_block_checksums: [data_block_count_max]u128,
+    data_block_indexes:   [data_block_count_max]u64,
+    data_block_key_max:   [data_block_count_max]Key,
+};
+
+const TableInfo = struct {
+    tree_id: u16,
+    index_block_index: u64,
+    index_block_checksum: u128,
+    key_min: Key,
+    key_max: Key,
+};
+```
+
+To lookup a value in a table, binary search the index block to locate the data block which should
+hold the value, then binary search inside the data block.
+
+Table size is physically limited by a single index block which can hold only so many references to
+data blocks. However, tables are further artificially limited to hold only a certain (compile-time
+constant) number of entries. Tables are arranged in levels. Each subsequent level contains
+exponentially more tables.
+
+Tables in a single level are pairwise disjoint. Tables in different layers overlap, but the key LSM:
+invariant is observed: values in shallow layers override values in deeper layers. This means that
+all modification happen to the first (purely in-memory) level.
+
+An asynchronous compaction process rebalances layers. Compaction removes one table from level A, finds
```

Comment:
`rebalances` needs be added to `scripts/.cspell.json` for the lint to pass. After that this LGTM!